agent/wsclient/client.go

```diff
@@ -290,75 +328,7 @@ func (cs *ClientServerImpl) handleMessage(data []byte) {
 	}
 }
 
-// websocketConn establishes a connection to the given URL, respecting any proxy configuration in the environment.
-// A standard proxying setup involves setting the following environment variables
-// (may be listed in /etc/ecs/ecs.config if using ecs-init):
-// HTTP_PROXY=http://<your-proxy>/ # HTTPS_PROXY may be set instead or additionally
-// NO_PROXY=169.254.169.254,/var/run/docker.sock # Directly connect to metadata service and docker socket
-func (cs *ClientServerImpl) websocketConn(parsedURL *url.URL, request *http.Request) (net.Conn, error) {
-	proxyURL, err := http.ProxyFromEnvironment(request)
-	if err != nil {
-		return nil, err
-	}
-
-	// url.Host might not have the port, but tls.Dial needs it
-	targetHost := parsedURL.Host
-	if !strings.Contains(targetHost, ":") {
-		targetHost += ":443"
-	}
-	targetHostname, _, err := net.SplitHostPort(targetHost)
-	if err != nil {
-		return nil, err
-	}
-
-	tlsConfig := tls.Config{ServerName: targetHostname, InsecureSkipVerify: cs.AcceptInvalidCert}
-	timeoutDialer := &net.Dialer{Timeout: wsConnectTimeout}
-
-	if proxyURL == nil {
-		// directly connect
-		seelog.Infof("Creating poll dialer, host: %s", parsedURL.Host)
-		return tls.DialWithDialer(timeoutDialer, "tcp", targetHost, &tlsConfig)
-	}
-
-	// connect via proxy
-	seelog.Infof("Creating poll dialer, proxy: %s", proxyURL.Host)
-	plainConn, err := timeoutDialer.Dial("tcp", proxyURL.Host)
-	if err != nil {
-		return nil, err
-	}
-
-	// TLS over an HTTP proxy via CONNECT taken from: https://golang.org/src/net/http/transport.go
-	connectReq := &http.Request{
-		Method: "CONNECT",
-		URL:    &url.URL{Opaque: targetHost},
-		Host:   targetHost,
-		Header: make(http.Header),
-	}
-
-	if proxyUser := proxyURL.User; proxyUser != nil {
-		username := proxyUser.Username()
-		password, _ := proxyUser.Password()
-		auth := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
-		connectReq.Header.Set("Proxy-Authorization", "Basic "+auth)
-	}
-
-	connectReq.Write(plainConn)
-
-	// Read response.
-	// Okay to use and discard buffered reader here, because
-	// TLS server will not speak until spoken to.
-	br := bufio.NewReader(plainConn)
-	resp, err := http.ReadResponse(br, connectReq)
-	if err != nil {
-		plainConn.Close()
-		return nil, err
-	}
-	if resp.StatusCode != 200 {
-		plainConn.Close()
-		return nil, fmt.Errorf(resp.Status)
-	}
-
-	tlsConn := tls.Client(plainConn, &tlsConfig)
-
-	return tlsConn, nil
+// See https://github.com/gorilla/websocket/blob/87f6f6a22ebfbc3f89b9ccdc7fddd1b914c095f9/conn.go#L650
+func permissibleCloseCode(err error) bool {
+	return websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway)
```

Comment:
I think we should add `CloseInternalServerErr` to this list as well. Also, this method needs a unit test.